### ISSUE TITLE
docker build --no-cache invalidates older downstream caches

### ISSUE THREAD

**Issue by xellsys** [moby/moby#39164]
#### tl;dr
With every run of `docker build --no-cache .` _all layers_ get new hashes - also if there are no changes in between builds. I suppose this leads to invalidating later caches if there is a "cache fork" (see examples below). Is this wanted behavior?

## Issue
I have noticed this behavior in our multi stage builds and broke it down to the following reproducible example:

#### multi stage
1. Create an onbuild image with an ARG:
```
# Dockerfile.onbuild
FROM alpine
ONBUILD ARG DATA
ONBUILD RUN echo "$DATA"
```
2. Build it:
`docker build -f Dockerfile.onbuild -t onbuild .`
3. Create two dockerfiles extending from that onbuild image:
```
# Dockerfile1
ARG DATA=data1
FROM onbuild
```
```
# Dockerfile2
ARG DATA=data2
FROM onbuild
```
4. Build the first Dockerfile:
`docker build -f Dockerfile1 .`
5. Build the first Dockerfile again, just to be sure the cache is being used:
`docker build -f Dockerfile1 .`
> Sending build context to Docker daemon   7.68kB
> Step 1/2 : ARG DATA=data1
> Step 2/2 : FROM onbuild
> \# Executing 2 build triggers
>  ---> Using cache
>  ---> Using cache
>  ---> 2c4d795c60c0
> Successfully built 2c4d795c60c0
6. Build the second Dockerfile without using the cache:
`docker build -f Dockerfile2 --no-cache .`
> Sending build context to Docker daemon   7.68kB
> Step 1/2 : ARG DATA=data2
> Step 2/2 : FROM onbuild
> \# Executing 2 build triggers
>  ---> Running in 6c4f6b640859
> Removing intermediate container 6c4f6b640859
>  ---> Running in 49fd7da24855
> data2
> Removing intermediate container 49fd7da24855
>  ---> 5b283f3cb437
> Successfully built 5b283f3cb437
7. Build the first Dockerfile using the cache again:
`docker build -f Dockerfile1 .`
* The build is not using the cache from the layer on where the two dockerfiles differ:
> Sending build context to Docker daemon   7.68kB
> Step 1/2 : ARG DATA=data1
> Step 2/2 : FROM onbuild
> \# Executing 2 build triggers
>  ---> Using cache
>  ---> Running in 7a4eeec3adfb
> data1
> Removing intermediate container 7a4eeec3adfb
>  ---> a5f1093b6920
> Successfully built a5f1093b6920

I would have expected that build Dockerfile2 would not have an effect on the build caches for Dockerfile1.
I can tell that a --no-cache build actually results in different layer hashes every time. This seems to be the issue here: _The Dockerfile2 build changes the hash for the first layer. The second layer created by the first Dockerfile1 build is not deriving from that new hash and thus needs to be rebuild._
In our scenario we have many builds extending from the same onbuild base image. If I rebuild one of those image with --no-cache I lose the benefit of the downstream caches for all the other image builds afterwards (from that point on where the execution actually differs from the initial --no-cache build).

#### non multi stage
I was also able to reproduce it outside of a multi stage context:
```
# Dockerfile
FROM alpine
RUN echo "first line, is always being cached"
COPY data .
RUN cat data
```
Run:
`echo "data1" > data`
`docker build .`
* now we change the data

`echo "data2" > data`
* and run a build without cache

`docker build --no-cache .`
* change data back to before

`echo "data1" > data`
`docker build .`
* not using the cache from the first build
